### ISSUE TITLE
"The connection to the server has been lost." Error of 2017-11-26.cbf1e48 Test Build

### ISSUE THREAD

**Issue by EnLightning** [freeorion/freeorion#1898]
### Environment

* **FreeOrion Version**:
v0.4.7+[build 2017-11-26.cbf1e48] MSVC 2015
* **Operating System**:
Windows 7 Pro
* **Graphic card used**:
NVIDIA GeForce 940MX
* **Fetched as** 
  * Weekly development build

### Description

Server crashes when advance a turn, the game exits to the main menu with pop-up "The connection to the server has been lost."

Error msgs from freeoriond.log

> 10:55:43.623500 [error] server : ObjectMap.cpp:128 : ObjectMap::FindObjects couldn't find object with id 194962761
10:55:43.623500 [error] server : ObjectMap.cpp:128 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.623500 [error] server : ObjectMap.cpp:128 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.626500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457666 m_zero = 0
10:55:43.626500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457666 is lower than the expected minimum new id this turn 0
10:55:43.626500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457666 is invalid.
10:55:43.627500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457658 m_zero = 0
10:55:43.627500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457658 is lower than the expected minimum new id this turn 0
10:55:43.627500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457658 is invalid.
10:55:43.627500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.627500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.629500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457650 m_zero = 0
10:55:43.629500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457650 is lower than the expected minimum new id this turn 0
10:55:43.629500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457650 is invalid.
10:55:43.629500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457642 m_zero = 0
10:55:43.629500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457642 is lower than the expected minimum new id this turn 0
10:55:43.629500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457642 is invalid.
10:55:43.629500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.629500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.629500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457634 m_zero = 0
10:55:43.629500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457634 is lower than the expected minimum new id this turn 0
10:55:43.629500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457634 is invalid.
10:55:43.629500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.630500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.630500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457626 m_zero = 0
10:55:43.630500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457626 is lower than the expected minimum new id this turn 0
10:55:43.630500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457626 is invalid.
10:55:43.630500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.630500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.632500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457618 m_zero = 0
10:55:43.632500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457618 is lower than the expected minimum new id this turn 0
10:55:43.632500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457618 is invalid.
10:55:43.632500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457610 m_zero = 0
10:55:43.632500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457610 is lower than the expected minimum new id this turn 0
10:55:43.632500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457610 is invalid.
10:55:43.632500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457602 m_zero = 0
10:55:43.632500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457602 is lower than the expected minimum new id this turn 0
10:55:43.632500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457602 is invalid.
10:55:43.632500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.632500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.633500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457594 m_zero = 0
10:55:43.633500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457594 is lower than the expected minimum new id this turn 0
10:55:43.633500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457594 is invalid.
10:55:43.633500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457586 m_zero = 0
10:55:43.633500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457586 is lower than the expected minimum new id this turn 0
10:55:43.633500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457586 is invalid.
10:55:43.633500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.633500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.634500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457578 m_zero = 0
10:55:43.634500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457578 is lower than the expected minimum new id this turn 0
10:55:43.634500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457578 is invalid.
10:55:43.639500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.639500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.639500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457570 m_zero = 0
10:55:43.639500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457570 is lower than the expected minimum new id this turn 0
10:55:43.639500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457570 is invalid.
10:55:43.639500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.639500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.639500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.639500 [error] server : IDAllocator.cpp:68 : m_empire_id -1 does not match apparent assiging id 2066705510 for id = -2143457562 m_zero = 0
10:55:43.639500 [error] server : IDAllocator.cpp:108 : invalid id = -2143457562 is lower than the expected minimum new id this turn 0
10:55:43.639500 [error] server : Universe.cpp:406 : An object has not been inserted into the universe because it's id = -2143457562 is invalid.
10:55:43.639500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.639500 [warn] server : Fleet.cpp:650 : Object map is missing ship with expected id -1
10:55:43.653500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1
10:55:43.653500 [error] server : ObjectMap.cpp:108 : ObjectMap::FindObjects couldn't find object with id -1

### Steps to reproduce

1) Download and unpack [Turn#111 save, config, and log files](https://github.com/freeorion/freeorion/files/1522917/111.zip) , get the save file 111.sav and load it.
2) Press button "Turn 111" to advance a turn.
3) The game exits to the main menu with pop-up "The connection to the server has been lost."

**Comment by dbenage-cx** [comment]
@EnLightning Thank you for report, any chance you have save file for previous turn? (And/or logs for turn 111 with some previous turn history)

Did the game initially crash when playing from a previous turn, or only after loading the save?

The existing save file crashes on load for me. If you do happen to have earlier save file, please resave after unchecking `Options`->`Other`->`Use Compression for XML Saves`

Suspect might related to #1865

**Comment by EnLightning** [comment]
This crash only happens on turn 111, I also reported [this issue in freeorion forum](http://www.freeorion.org/forum/viewtopic.php?f=28&t=10807),  @Vezzra said:
> Tested your save on OSX and Win7. On OSX, the game crashes immediately (with "Connection to server lost") upon loading, on Win7 I can load, but when I try to advance a turn, turn processing apparently hangs at the fleet movement phase. When I finally resigned, I got the "Connection to server lost" error.

Here I attached the uncompressed save files of recent 4 turns for you guys to debug.

[108-111.zip](https://github.com/freeorion/freeorion/files/1523339/108-111.zip)

**Comment by dbenage-cx** [comment]
On load of save from turn 108:
```
19:47:23.415362 [error] server : ObjectMap.cpp:118 : ObjectMap::FindObjects couldn't find object with id 59808269
19:47:23.415481 [error] server : ObjectMap.cpp:118 : ObjectMap::FindObjects couldn't find object with id 273760426
```
Both ids are fleets for AI_6, which are now empty/destroyed.

Clicking next turn has same result as #1865 (I'm unclear how we end up with invalid pointer here).
Leaving this open and title unmodified since the issues may have different root causes.

**Comment by Vezzra** [comment]
With the uncompressed save files, I can load the game on OSX without getting an immediate crash. However, only when I load the savegame for turn 111, I get to try to advance to turn 112 (which results in a crash of course).

When I load the savegame turn 110, I already get the crash when advancing to turn 111. Same happens when I load the savegame for turn 108: I can advance until 110 just fine, when trying to advance to 111 -> crash.

**Comment by dbenage-cx** [comment]
Applying either of these diffs prevents previous noted crashes for me (hitting next turn on any of the above attached saves).

Store the unfiltered `m_offset_to_empire_id`:
```diff
diff --git a/universe/IDAllocator.cpp b/universe/IDAllocator.cpp
index 5592e839b..92bc448c2 100644
--- a/universe/IDAllocator.cpp
+++ b/universe/IDAllocator.cpp
@@ -309,7 +309,7 @@ void IDAllocator::SerializeForEmpire(Archive& ar, const unsigned int version, in
             // Filter the map for empires so they only have their own actual next id and no
             // information about other clients.
             std::unordered_map<int, int> temp_empire_id_to_object_id{};
-            auto temp_offset_to_empire_id = std::vector<int>(m_offset_to_empire_id.size(), m_server_id);
+            //auto temp_offset_to_empire_id = std::vector<int>(m_offset_to_empire_id.size(), m_server_id);
 
             auto&& it = m_empire_id_to_next_assigned_object_id.find(empire_id);
             if (it == m_empire_id_to_next_assigned_object_id.end()) {
@@ -317,11 +317,11 @@ void IDAllocator::SerializeForEmpire(Archive& ar, const unsigned int version, in
                               << empire_id << " not in id manager table.";
             } else {
                 temp_empire_id_to_object_id.insert(*it);
-                temp_offset_to_empire_id[it->second % m_stride] = empire_id;
+                //temp_offset_to_empire_id[it->second % m_stride] = empire_id;
             }
 
             ar & boost::serialization::make_nvp(BOOST_PP_STRINGIZE(m_empire_id_to_next_assigned_object_id), temp_empire_id_to_object_id);
-            ar & boost::serialization::make_nvp(BOOST_PP_STRINGIZE(m_offset_to_empire_id), temp_offset_to_empire_id);
+            ar & boost::serialization::make_nvp(BOOST_PP_STRINGIZE(m_offset_to_empire_id), m_offset_to_empire_id);
 
             DebugLogger(IDallocator) << "Serialized [" << [this, &temp_empire_id_to_object_id](){
                 std::stringstream ss;
```

Prevent out of bounds index in `temp_offset_to_empire_id`:
```diff
diff --git a/universe/IDAllocator.cpp b/universe/IDAllocator.cpp
index 5592e839b..d9ce38537 100644
--- a/universe/IDAllocator.cpp
+++ b/universe/IDAllocator.cpp
@@ -317,7 +317,13 @@ void IDAllocator::SerializeForEmpire(Archive& ar, const unsigned int version, in
                               << empire_id << " not in id manager table.";
             } else {
                 temp_empire_id_to_object_id.insert(*it);
-                temp_offset_to_empire_id[it->second % m_stride] = empire_id;
+                std::size_t offset_idx  = it->second % m_stride;
+                if (temp_offset_to_empire_id.size() <= offset_idx) {
+                    ErrorLogger(IDallocator) << "Invalid index " << offset_idx << " from next id " << it->second
+                                             << " with stride " << m_stride;
+                } else {
+                    temp_offset_to_empire_id.at(offset_idx) = empire_id;
+                }
             }
 
             ar & boost::serialization::make_nvp(BOOST_PP_STRINGIZE(m_empire_id_to_next_assigned_object_id), temp_empire_id_to_object_id);
```

Log excerpt with second diff:
`[error] IDallocator : IDAllocator.cpp:322 : Invalid index 18446744073709551610 from next id -2108436422 with stride 8`

**Comment by Vezzra** [comment]
Tested both patches on OSX, and both work. Tested with the savegames of turn 108 and 111, and could advance to turn 113 without issues in all cases.

@dbenage-cx, can't tell which of these patches makes more sense, but if one of them produces errors in the logs and the other doesn't, then the no-error-producing one is probably preferable.

So, as far as I'm concerned, go ahead and put the fix in.

**Comment by geoffthemedio** [comment]
Does this have any interactions with https://github.com/freeorion/freeorion/pull/1903 ?

**Comment by LGM-Doyle** [comment]
@dbenage-cx, perusing the loaded state of the game, the state is already corrupted.  

i think that the underlying problem is twofold: the serialization of `IDAllocator` skips `m_zero` and the line `it->second % m_stride` should be `(it->second -m_zero) % m_stride`.  Depending on how particular compilers and versions of boost serialization defaulted `m_zero`, the `IDAllocator` state could end up wildly incorrect.

[Fix IDAllocator missing m_zero in its serialization](https://github.com/freeorion/freeorion/pull/1903) fixes these problems.

The first fix here discards the filtering of `m_offset_to_empire_id` and would result in empires being able to create objects/designs for other empires.

In the second case the fix will prevent some empires from creating any ship designs, since `offset_to_empire_at_id` does not contain any client's empire id.  However, adding check to deserialization that all next ids are greater than `m_zero`,  less than `m_warn_threshold` and `m_exhausted_threshold` and of the expected modulus is a good idea.  It would have prevented the crash and provided a clear error message in this case even if game function would be impaired.  Would you like to implement this?

*(Edit by Vezzra: fixed link for "Fix IDAllocator missing m_zero in its serialization")*

**Comment by dbenage-cx** [comment]
> Would you like to implement this?

@LGM-Doyle Unfortunately, I do not grok the implementation of `IDallocator` enough (yet).
If you find time and motivation to implement, I'd not wish to postpone for me to catch up to speed here.
I will commit to understanding it better in either case.

With exception for special casing, I assume a negative id is from overflow and should be invalid, but can not attest to where such issue arises (or ascertain relevance wrt #1903).
I'd suspect a point at one or more of:
*  Prior to serialization
*  During deserialization on originating computer (or platform/compiler/dependencies)
*  During deserialization on a different computer (or platform/compiler/dependencies)

Thus my reasoning for requesting an earlier save file. I've only found a negative value of `-32768` so far in the earliest save (mostly some tech and turn ids), so not determined if "cross-platform" is an issue or if it simply masks the issue.
My current feeling is that #1903 will address the underlying initial problem, with some lingering issue for some save games between platforms(/etc...).

@Vezzra Apologies for time wasted testing the above patches.  I should have noted they were intended to help facilitate further troubleshooting.

OT: Should a mandatory issue be opened for the note at [IDallocator.h:35](https://github.com/freeorion/freeorion/blob/master/universe/IDAllocator.h#L35)?
(`TODO: Remove the extra code legacy loading before v0.4.8`)

**Comment by LGM-Doyle** [comment]
@dbenage-cx I agree with your reasoning.  Finding the save game the turn that the corruption starts is the best way to determine the cause.  I extended the max number of autosaved games to 10000, in #1906,  to make it more likely that testers and devs have the saved game on hand.

When I produced #1903, I was not actually looking at this issue.  I was trying to determine the cause of the "m_empire_id X does not match apparent assigning id" errors.  There may be further problems.

I've created the issue https://github.com/freeorion/freeorion/issues/1909

**Comment by Vezzra** [comment]
> @Vezzra Apologies for time wasted testing the above patches. I should have noted they were intended to help facilitate further troubleshooting.

No problem. These things happen, and it hasn't really taken that much time to run those brief tests.

**Comment by Vezzra** [comment]
@EnLightning, why did you close this? Is the issue fixed for you in the recent test builds? (Just asking to make sure.)

**Comment by EnLightning** [comment]
@Vezzra Yes, I've tried recent 3 weekly releases, never encountered it.

**Comment by Vezzra** [comment]
Perfect, thanks!
